spikeinterface/sortingcomponents/peak_detection.py

```diff
@@ -21,6 +21,116 @@
+
+    """Method to subsample all the found peaks before clustering
+
+    Parameters
+    ----------
+    peaks: the peaks that have been found
+    method: 'random', 'smart_sampling'
+        Method to use. Options:
+            * 'random' : a random subset is select from all the peaks
+            * 'smart_sampling' : peaks are selected by monte-carlo rejections
+    max_peaks_per_channel: int
+        The maximal number of peaks that should be kept per channel.
+    detect_threshold: float
+        Threshold, in median absolute deviations (MAD), to use to detect peaks.
+    seed: int
+        The seed for random generations
+    method_kwargs : dict
+        If the methods have particular parameters, they can be provided here
+    {}
+
+    Returns
+    -------
+    peaks: array
+        Detected peaks.
+
+    Notes
+    -----
+    This peak detection ported from tridesclous into spikeinterface.
+    """
+
+    selected_peaks = []
+    peaks_indices = {}
+
+    if seed is not None:
+        np.random.seed(seed)
+
+    for channel in np.unique(peaks['channel_ind']):
+        peaks_indices[channel] = np.where(peaks['channel_ind'] == channel)[0]
+
+    if method == 'random':
+        for channel in np.unique(peaks['channel_ind']):
+            selected_peaks += [np.random.permutation(peaks_indices[channel])[:max_peaks_per_channel]]
+    elif method == 'smart_sampling':
+
+        params = {'detect_threshold' : 5, 
+                  'peak_sign' : 'neg',
+                  'n_bins' : 50}
+
+        params.update(method_kwargs)
+        assert 'noise_levels' in params
+
+        abs_threholds = params['noise_levels']*params['detect_threshold']
+
+        import scipy
+
+        def reject_rate(x, d, a, target, n_bins):
+            return (np.mean(n_bins*a*np.clip(1 - d*x, 0, 1)) - target)**2
+
+        histograms = {}
+        for channel in np.unique(peaks['channel_ind']):
+
+            sub_peaks = peaks[peaks_indices[channel]]
+
+            if params['peak_sign'] == 'neg':
+                bins = list(np.linspace(sub_peaks['amplitude'].min(), -abs_threholds[channel], params['n_bins']))
+            elif params['peak_sign'] == 'pos':
+                bins = list(abs_threholds[channel], np.linspace(sub_peaks['amplitude'].max(), params['n_bins']))
+            elif params['peak_sign'] == 'both':
+                if sub_peaks['amplitude'].max() > abs_threholds[channel]:
+                    pos_values = list(abs_threholds[channel], np.linspace(sub_peaks['amplitude'].max(), params['n_bins']//2))
+                else:
+                    pos_values = []
+                if sub_peaks['amplitude'].min() < -abs_threholds[channel]:
+                    neg_values = list(np.linspace(sub_peaks['amplitude'].min(), -abs_threholds[channel], params['n_bins']//2))
+                else:
+                    neg_values = []
+                bins = neg_values + pos_values
+
```

Comment:
You should add a small comments that explain the procedure in 2 or 3 lines here.

Reply:
done